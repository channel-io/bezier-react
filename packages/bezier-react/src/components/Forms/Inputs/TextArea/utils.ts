import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
