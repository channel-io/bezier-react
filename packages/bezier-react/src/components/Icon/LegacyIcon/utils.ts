/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
