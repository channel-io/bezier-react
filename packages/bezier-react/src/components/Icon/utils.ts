/* Internal Dependencies */
import icons, { IconName } from './generated'

export const isIconName = (name): name is IconName => Object.keys(icons).includes(name)
