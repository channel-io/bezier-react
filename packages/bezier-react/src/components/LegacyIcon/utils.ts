import {
  type IconName,
  icons,
} from '@channel.io/bezier-icons'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { AllIcon } from '@channel.io/bezier-icons'
 * import { Icon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-icons'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(() => <svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
