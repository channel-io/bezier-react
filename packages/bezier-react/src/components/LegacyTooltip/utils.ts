import {
  type GetReplacement,
  type GetTooltipStyle,
  LegacyTooltipPosition,
} from './LegacyTooltip.types'

export function getReplacement({
  tooltip,
  keepInContainer,
  placement,
  rootElement,
}: GetReplacement): LegacyTooltipPosition {
  if (!keepInContainer) {
    return placement
  }

  const {
    width: tooltipWidth,
    height: tooltipHeight,
    top: tooltipTop,
    left: tooltipLeft,
  } = tooltip.getBoundingClientRect()

  const {
    width: rootWidth,
    height: rootHeight,
    top: rootTop,
    left: rootLeft,
  } = rootElement.getBoundingClientRect()

  const isOverTop = tooltipTop < rootTop
  const isOverBottom = tooltipTop + tooltipHeight > rootTop + rootHeight
  const isOverLeft = tooltipLeft < rootLeft
  const isOverRight = tooltipLeft + tooltipWidth > rootLeft + rootWidth

  if (isOverTop) {
    if (isOverLeft) {
      return LegacyTooltipPosition.BottomLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.BottomRight
    }
    return LegacyTooltipPosition.BottomCenter
  }

  if (isOverBottom) {
    if (isOverLeft) {
      return LegacyTooltipPosition.TopLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.TopRight
    }
    return LegacyTooltipPosition.TopCenter
  }

  if (isOverLeft) {
    if (isOverTop) {
      return LegacyTooltipPosition.RightTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.RightBottom
    }
    return LegacyTooltipPosition.RightCenter
  }

  if (isOverRight) {
    if (isOverTop) {
      return LegacyTooltipPosition.LeftTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.LeftBottom
    }
    return LegacyTooltipPosition.LeftCenter
  }

  return placement
}

export function getTooltipStyle({
  tooltipContainer,
  placement,
  offset,
  allowHover,
}: GetTooltipStyle) {
  const {
    top: containerTop,
    left: containerLeft,
    width: containerWidth,
    height: containerHeight,
  } = tooltipContainer.getBoundingClientRect()

  let top = containerTop
  let left = containerLeft
  let translateX = 0
  let translateY = 0

  let offsetTop = 0
  let offsetRight = 0
  let offsetBottom = 0
  let offsetLeft = 0
  let offsetStyle = {}

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.TopLeft:
    case LegacyTooltipPosition.TopRight:
      translateY = -100
      offsetBottom = offset
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.RightTop:
    case LegacyTooltipPosition.RightBottom:
      left += containerWidth
      offsetLeft = offset
      break
    case LegacyTooltipPosition.BottomCenter:
    case LegacyTooltipPosition.BottomLeft:
    case LegacyTooltipPosition.BottomRight:
      top += containerHeight
      offsetTop = offset
      break
    case LegacyTooltipPosition.LeftCenter:
    case LegacyTooltipPosition.LeftTop:
    case LegacyTooltipPosition.LeftBottom:
      translateX = -100
      offsetRight = offset
      break
  }

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.BottomCenter:
      translateX = -50
      left += (containerWidth / 2)
      break
    case LegacyTooltipPosition.TopRight:
    case LegacyTooltipPosition.BottomRight:
      translateX = -100
      left += containerWidth
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.LeftCenter:
      translateY = -50
      top += (containerHeight / 2)
      break
    case LegacyTooltipPosition.RightBottom:
    case LegacyTooltipPosition.LeftBottom:
      translateY = -100
      top += containerHeight
      break
  }

  if (allowHover) {
    offsetStyle = {
      paddingTop: offsetTop,
      paddingRight: offsetRight,
      paddingBottom: offsetBottom,
      paddingLeft: offsetLeft,
    }
  } else {
    top += (offsetTop - offsetBottom)
    left += (offsetLeft - offsetRight)
  }

  return {
    position: 'fixed',
    top,
    left,
    transform: `translate(${translateX}%, ${translateY}%)`,
    ...offsetStyle,
  }
}
