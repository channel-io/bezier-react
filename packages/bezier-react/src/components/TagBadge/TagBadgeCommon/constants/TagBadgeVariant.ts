enum TagBadgeVariant {
  Default = 'Default',
  MonochromeLight = 'MonochromeLight',
  MonochromeDark = 'MonochromeDark',
  Blue = 'Blue',
  Cobalt = 'Cobalt',
  Teal = 'Teal',
  Green = 'Green',
  Olive = 'Olive',
  Pink = 'Pink',
  Navy = 'Navy',
  Yellow = 'Yellow',
  Orange = 'Orange',
  Red = 'Red',
  Purple = 'Purple',
}

export const TagBadgeBgColorPreset = {
  [TagBadgeVariant.Default]: 'bg-black-lighter',
  [TagBadgeVariant.MonochromeLight]: 'bg-black-lighter',
  [TagBadgeVariant.MonochromeDark]: 'bg-black-darker',
  [TagBadgeVariant.Blue]: 'bgtxt-blue-lighter',
  [TagBadgeVariant.Cobalt]: 'bgtxt-cobalt-lighter',
  [TagBadgeVariant.Teal]: 'bgtxt-teal-lighter',
  [TagBadgeVariant.Green]: 'bgtxt-green-lighter',
  [TagBadgeVariant.Olive]: 'bgtxt-olive-lighter',
  [TagBadgeVariant.Pink]: 'bgtxt-pink-lighter',
  [TagBadgeVariant.Navy]: 'bgtxt-navy-lighter',
  [TagBadgeVariant.Yellow]: 'bgtxt-yellow-lighter',
  [TagBadgeVariant.Orange]: 'bgtxt-orange-lighter',
  [TagBadgeVariant.Red]: 'bgtxt-red-lighter',
  [TagBadgeVariant.Purple]: 'bgtxt-purple-lighter',
} as const

export const BadgeColorPreset = {
  [TagBadgeVariant.Default]: 'txt-black-darkest',
  [TagBadgeVariant.MonochromeLight]: 'txt-black-dark',
  [TagBadgeVariant.MonochromeDark]: 'bgtxt-absolute-white-dark',
  [TagBadgeVariant.Blue]: 'bgtxt-blue-normal',
  [TagBadgeVariant.Cobalt]: 'bgtxt-cobalt-normal',
  [TagBadgeVariant.Teal]: 'bgtxt-teal-normal',
  [TagBadgeVariant.Green]: 'bgtxt-green-normal',
  [TagBadgeVariant.Olive]: 'bgtxt-olive-normal',
  [TagBadgeVariant.Pink]: 'bgtxt-pink-normal',
  [TagBadgeVariant.Navy]: 'bgtxt-navy-normal',
  [TagBadgeVariant.Yellow]: 'bgtxt-yellow-normal',
  [TagBadgeVariant.Orange]: 'bgtxt-orange-normal',
  [TagBadgeVariant.Red]: 'bgtxt-red-normal',
  [TagBadgeVariant.Purple]: 'bgtxt-purple-normal',
} as const

export default TagBadgeVariant
