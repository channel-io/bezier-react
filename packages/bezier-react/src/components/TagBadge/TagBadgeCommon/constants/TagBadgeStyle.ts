/* Internal dependencies */
import { IconSize } from 'Components/Icon'

export const BADGE_TEXT_HORIZONTAL_PADDING = 3

export const TAG_TEXT_HORIZONTAL_PADDING = 2

export const TAGBADGE_VERTICAL_PADDING = 1

export const TAG_BADGE_ICON_SIZE = IconSize.XS
