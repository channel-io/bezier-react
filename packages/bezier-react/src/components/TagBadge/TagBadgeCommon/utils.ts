import { type SemanticColor } from '~/src/types/tokens'

import TagBadgeSize from './constants/TagBadgeSize'
import type TagBadgeVariant from './constants/TagBadgeVariant'
import { TagBadgeBgColorPreset } from './constants/TagBadgeVariant'

export function getProperTagBadgeTypo(size: TagBadgeSize) {
  return {
    [TagBadgeSize.XS]: '11' as const,
    [TagBadgeSize.S]: '13' as const,
    [TagBadgeSize.M]: '14' as const,
    [TagBadgeSize.L]: '15' as const,
  }[size]
}

export function getProperTagBadgeBgColor(color: TagBadgeVariant): SemanticColor {
  return TagBadgeBgColorPreset[color]
}
