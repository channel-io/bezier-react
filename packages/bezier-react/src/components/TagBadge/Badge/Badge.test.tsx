/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { RoundAbsoluteNumber } from '~/src/foundation'
import { render } from '~/src/utils/testUtils'
import {
  TagBadgeSize,
  getProperTagBadgePadding,
} from '~/src/components/TagBadge/TagBadgeCommon'
import { TAGBADGE_VERTICAL_PADDING } from '~/src/components/TagBadge/TagBadgeCommon/constants/TagBadgeStyle'
import {
  Badge,
  BADGE_TEST_ID,
} from './Badge'
import type BadgeProps from './Badge.types'

describe('Badge test >', () => {
  let props: BadgeProps

  beforeEach(() => {
    props = {}
  })

  const renderBadge = (optionProps?: BadgeProps) => render(<Badge {...props} {...optionProps} />)

  it('Snapshot >', () => {
    const { getByTestId } = renderBadge()
    const badge = getByTestId(BADGE_TEST_ID)

    expect(badge).toMatchSnapshot()
  })

  describe('Size Test >', () => {
    it('xs', () => {
      const size = TagBadgeSize.XS
      const { getByTestId } = renderBadge({ size })
      const xsBadge = getByTestId(BADGE_TEST_ID)

      expect(xsBadge).toHaveStyle(`padding: ${TAGBADGE_VERTICAL_PADDING}px ${getProperTagBadgePadding(size)}px`)
      expect(xsBadge).toHaveStyle(`border-radius: ${RoundAbsoluteNumber.R4}px`)
    })

    it('s', () => {
      const size = TagBadgeSize.S
      const { getByTestId } = renderBadge({ size })
      const xBadge = getByTestId(BADGE_TEST_ID)

      expect(xBadge).toHaveStyle(`padding: ${TAGBADGE_VERTICAL_PADDING}px ${getProperTagBadgePadding(size)}px`)
      expect(xBadge).toHaveStyle(`border-radius: ${RoundAbsoluteNumber.R6}px`)
    })
    it('m', () => {
      const size = TagBadgeSize.M
      const { getByTestId } = renderBadge({ size })
      const mBadge = getByTestId(BADGE_TEST_ID)

      expect(mBadge).toHaveStyle(`padding: ${TAGBADGE_VERTICAL_PADDING}px ${getProperTagBadgePadding(size)}px`)
      expect(mBadge).toHaveStyle(`border-radius: ${RoundAbsoluteNumber.R6}px`)
    })
    it('l', () => {
      const size = TagBadgeSize.L
      const { getByTestId } = renderBadge({ size })
      const lBadge = getByTestId(BADGE_TEST_ID)

      expect(lBadge).toHaveStyle(`padding: ${TAGBADGE_VERTICAL_PADDING}px ${getProperTagBadgePadding(size)}px`)
      expect(lBadge).toHaveStyle(`border-radius: ${RoundAbsoluteNumber.R6}px`)
    })
  })
})
