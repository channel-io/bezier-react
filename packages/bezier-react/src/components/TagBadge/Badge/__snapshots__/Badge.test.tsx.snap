// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge test > Snapshot > 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  border-radius: 6px;
}

<div
  class="c0"
  color="txt-black-darkest"
  data-testid="bezier-react-badge"
  style="padding: 1px 4px; color: rgba(0, 0, 0, 0.851); background-color: rgba(0, 0, 0, 0.051);"
/>
`;
